subsys/power/rt_dpm.c

```diff
@@ -0,0 +1,304 @@
+/*
+ * Copyright (c) 2020 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <ksched.h>
+#include <wait_q.h>
+#include <spinlock.h>
+#include <power/rt_dpm.h>
+
+/* States defined for device runtime power management.
+ */
+enum rt_dpm_state {
+	RT_DPM_ACTIVE,
+	RT_DPM_RESUMING,
+	RT_DPM_SUSPENDED,
+	RT_DPM_SUSPENDING
+};
+
+/* Try to suspend the given device, if successfully make
+ * the state transition, asynchronously release its parent
+ * to avoid holding caller context long time.
+ */
+static int rt_dpm_release_helper(struct rt_dpm *rt_pm)
+{
+	int ret = 0;
+	k_spinlock_key_t key;
+	struct k_thread *thread;
+	struct device *dev = CONTAINER_OF(rt_pm, struct device, rt_pm);
+
+	key = k_spin_lock(&rt_pm->lock);
+
+	if (rt_pm->disable_count > 0) {
+		k_spin_unlock(&rt_pm->lock, key);
+		return -EACCES;
+	} else if (rt_pm->usage_count > 0 ||
+		   rt_pm->state == RT_DPM_SUSPENDED) {
+		k_spin_unlock(&rt_pm->lock, key);
+		return 0;
+	}
+
+	rt_pm->state = RT_DPM_SUSPENDING;
+
+	if (rt_pm->ops && rt_pm->ops->pre_suspend) {
+		k_spin_unlock(&rt_pm->lock, key);
+		ret = (rt_pm->ops->pre_suspend)(dev);
+		key = k_spin_lock(&rt_pm->lock);
+	}
+
+	if (ret) {
+		rt_pm->state = RT_DPM_ACTIVE;
+	} else {
+		if (rt_pm->ops && rt_pm->ops->suspend) {
+			(rt_pm->ops->suspend)(dev);
+		}
+		rt_pm->state = RT_DPM_SUSPENDED;
+	}
+
+	/* resume the pended thread for the possible concurrency */
+	thread = z_unpend_first_thread(&rt_pm->wait_q);
```

Comment:
we should consider using condition variables here, I think they might give you what you want in a generic way, see 

https://github.com/nashif/zephyr/pull/new/condition_variables_01 

I created a very quick/dirst sample for PM to illustrate how they can be used.